meson.build

```diff
@@ -0,0 +1,350 @@
+        'warning_level=3',
+    ],
+    meson_version: '>= 1.1.0',
+)
+
+v_array = meson.project_version().split('.')
+v_maj = v_array[0]
+
+# config.h
+
+config_dir = [include_directories('.')]
+
+config_h = configuration_data()
+config_h.set_quoted('PACKAGE', 'xz')
+config_h.set_quoted('PACKAGE_BUGREPORT', 'xz@tukaani.org')
+config_h.set_quoted('PACKAGE_NAME', meson.project_name())
+config_h.set_quoted('PACKAGE_URL', 'https://tukaani.org/xz/')
+
+# host
+
+config_h.set10('WORDS_BIGENDIAN', host_machine.endian() == 'big',
+    description: 'Define to 1 if the processor stores words with the most significant byte first.'
+)
+
+host_os = host_machine.system()
+
+cygwin = 'cygwin'
+windows = 'windows'
+linux = 'linux'
+sunos = 'sunos'
+asm_os = [ 'linux', 'dragonfly', 'freebsd', 'netbsd', 'openbsd', 'windows', 'cygwin' ]
+sys_cygwin = cygwin.contains(host_os)
+sys_windows = windows.contains(host_os)
+
+COND_W32 = sys_cygwin or sys_windows
+
+# binaries
+cc = meson.get_compiler('c')
+
+# try mimic AC_USE_SYSTEM_EXTENSIONS
+if host_machine.system() == 'linux'
+    add_project_arguments('-D_GNU_SOURCE', language: 'c')
+elif host_machine.system() == 'sunos'
+    add_project_arguments('-D__EXTENSIONS__', language: 'c')
+    add_project_arguments('-D_POSIX_PTHREAD_SEMANTICS', language: 'c')
+else
+    add_project_arguments('-D_TANDEM_SOURCE', language: 'c')
+    add_project_arguments('-D_ALL_SOURCE', language: 'c')
+endif
+
+lzma_c_args = []
+cflags_try = []
+
+if cc.get_argument_syntax() == 'gcc'
+    cflags_try += [
+        '-Wvla',
+        '-Wformat=2',
+        '-Winit-self',
+        '-Wmissing-include-dirs',
+        '-Wshift-overflow=2',
+        '-Wstrict-overflow=3',
+        '-Walloc-zero',
+        '-Wduplicated-cond',
+        '-Wfloat-equal',
+        '-Wundef',
+        '-Wshadow',
+        '-Wpointer-arith',
+        '-Wbad-function-cast',
+        '-Wwrite-strings',
+        '-Wdate-time',
+        '-Wsign-conversion',
+        '-Wfloat-conversion',
+        '-Wlogical-op',
+        '-Waggregate-return',
+        '-Wstrict-prototypes',
+        '-Wold-style-definition',
+        '-Wmissing-prototypes',
+        '-Wmissing-declarations',
+        '-Wredundant-decls',
+#
+        '-Wc99-compat',
+        '-Wc11-extensions',
+        '-Wc2x-compat',
+        '-Wc2x-extensions',
+        '-Wpre-c2x-compat',
+        '-Warray-bounds-pointer-arithmetic',
+        '-Wassign-enum',
+        '-Wconditional-uninitialized',
+        '-Wdocumentation',
+        '-Wduplicate-enum',
+        '-Wempty-translation-unit',
+        '-Wflexible-array-extensions',
+        '-Wmissing-variable-declarations',
+        '-Wnewline-eof',
+        '-Wshift-sign-overflow',
+        '-Wstring-conversion',
+    ]
+
+    lzma_c_args += cc.get_supported_arguments(cflags_try)
+endif
+
+config_h.set10('HAVE_VISIBILITY', cc.has_function_attribute('visibility'),
+    description: 'Define to 1 if the compiler supports simple visibility declarations.'
+)
```

Comment:
Meson internally handles -Werror for exactly that reason:
.https://github.com/mesonbuild/meson/blob/7d28ff29396f9d7043204de8ddc52226b9903811/mesonbuild/compilers/mixins/clike.py#L1309-L1324

Most projects I have seen, actually check for `__GNUC__` instead of using visibility.m4 even with autotools. This is kind of necessary anyway since you need preprocessor logic to figure out whether Microsoft dllexport/dllimport is viable, and visibility.m4 is ultimately not all that helpful overall.


Certain versions of... SunPro, if I recall correctly, are the only commonly known compiler that handles visibility but isn't either `__GNUC__` or `MSC_VER`, and they use a third syntax instead?

Reply:
> Meson internally handles -Werror for exactly that reason:

Great! Documenting it could be nice to give more confidence. Or perhaps it's that coming from other build systems one is constantly suspecting that corner cases aren't handled well enough, sorry.

> Most projects I have seen, actually check for `__GNUC__` instead of using visibility.m4 even with autotools.

liblzma checks in this order in the C code:

1. `#if defined(_WIN32) || defined(__CYGWIN__)` then use `__declspec(dllexport)`.

2. `#if HAVE_VISIBILITY` to use GNU C attributes. Checking for `__GNUC__` wouldn't work as it can be defined on platforms that don't support the visibility attributes; one has to know if the attribute is actually supported.

3. Otherwise assume no visibility support.

> Certain versions of... SunPro, if I recall correctly, are the only commonly known compiler that handles visibility but isn't either `__GNUC__` or `MSC_VER`, and they use a third syntax instead?

My note in b0d1422b6037bfea6f6723683bd82a8e6d77026c says that SunPro (or whatever its name is this year) supports the GNU C attribute since 2007. At least modern releases support the `-fvisibility` option too. I didn't research when the command line option was added (it would make sense to add it when the attribute was but I don't know if that happened). The original option is `-xldscope` but I guess there's no point to add support for it when `-fvisibility` works too.

Reply:
This

```
__attribute__((dllexport)) int foo(int x) { return x; }
```

results in a warning with GCC 14.1.1 on x86-64 GNU/Linux:

```
warning: ‘dllexport’ attribute directive ignored [-Wattributes]
    1 | __attribute__((dllexport)) int foo(int x) { return x; }
      | ^~~~~~~~~~~~~
```

But `cc.has_function_attribute('dllexport')` is still true. So `-Werror` isn't used or Meson shortcircuits the attribute check somehow. Thus I fear it might do the same with the visibility attribute as well.